instrumentation/net/http/otelhttp/internal/semconv/v1.24.0.go

```diff
@@ -0,0 +1,156 @@
+	"slices"
+	"strings"
+
+	"go.opentelemetry.io/otel/attribute"
+	semconvNew "go.opentelemetry.io/otel/semconv/v1.24.0"
+)
+
+type newHTTPServer struct{}
+
+var _ HTTPServer = newHTTPServer{}
+
+// TraceRequest returns trace attributes for an HTTP request received by a
+// server.
+//
+// The server must be the primary server name if it is known. For example this
+// would be the ServerName directive
+// (https://httpd.apache.org/docs/2.4/mod/core.html#servername) for an Apache
+// server, and the server_name directive
+// (http://nginx.org/en/docs/http/ngx_http_core_module.html#server_name) for an
+// nginx server. More generically, the primary server name would be the host
+// header value that matches the default virtual host of an HTTP server. It
+// should include the host identifier and if a port is used to route to the
+// server that port identifier should be included as an appropriate port
+// suffix.
+//
+// If the primary server name is not known, server should be an empty string.
+// The req Host will be used to determine the server instead.
+func (n newHTTPServer) RequestTraceAttrs(server string, req *http.Request) []attribute.KeyValue {
+	const MaxAttributes = 11
+	attrs := make([]attribute.KeyValue, MaxAttributes)
+	var host string
+	var p int
+	if server == "" {
+		host, p = splitHostPort(req.Host)
+	} else {
+		// Prioritize the primary server name.
+		host, p = splitHostPort(server)
+		if p < 0 {
+			_, p = splitHostPort(req.Host)
+		}
+	}
+
+	attrs[0] = semconvNew.ServerAddress(host)
+	i := 1
+	if hostPort := requiredHTTPPort(req.TLS != nil, p); hostPort > 0 {
+		attrs[i] = semconvNew.ServerPort(hostPort)
+		i++
+	}
+	i += n.method(req.Method, attrs[i:])     // Max 2
+	i += n.scheme(req.TLS != nil, attrs[i:]) // Max 1
+
+	if peer, peerPort := splitHostPort(req.RemoteAddr); peer != "" {
+		// The Go HTTP server sets RemoteAddr to "IP:port", this will not be a
+		// file-path that would be interpreted with a sock family.
+		attrs[i] = semconvNew.NetworkPeerAddress(peer)
+		i++
+		if peerPort > 0 {
+			attrs[i] = semconvNew.NetworkPeerPort(peerPort)
+			i++
+		}
+	}
+
+	if useragent := req.UserAgent(); useragent != "" {
+		// This is the same between v1.20, and v1.24
+		attrs[i] = semconvNew.UserAgentOriginal(useragent)
+		i++
+	}
+
+	if clientIP := serverClientIP(req.Header.Get("X-Forwarded-For")); clientIP != "" {
+		attrs[i] = semconvNew.ClientAddress(clientIP)
+		i++
+	}
+
+	if req.URL != nil && req.URL.Path != "" {
+		attrs[i] = semconvNew.URLPath(req.URL.Path)
+		i++
+	}
+
+	protoName, protoVersion := netProtocol(req.Proto)
+	if protoName != "" && protoName != "http" {
+		attrs[i] = semconvNew.NetworkProtocolName(protoName)
+		i++
+	}
+	if protoVersion != "" {
+		attrs[i] = semconvNew.NetworkProtocolVersion(protoVersion)
+		i++
+	}
+
+	return slices.Clip(attrs[:i])
+}
+
+func (n newHTTPServer) method(method string, attrs []attribute.KeyValue) int {
+	if method == "" {
+		attrs[0] = semconvNew.HTTPRequestMethodGet
```

Comment:
It is not obvious that `attrs` passed here has to be pre-allocated.

Using an `append` approach would avoid the panic if `len(attrs) <= 0`.